### PR TITLE
CMake Build Refactor and Cleanup

### PR DESCRIPTION
**Pull request by vector-of-bool**
This PR cleans up and modularizes the CMake build, and now permits building with Visual C++ on Windows via CMake.

### CODE REVIEW

File: test/tests/CMakeLists.txt [hunk 1]
```diff
@@ -31,7 +31,9 @@ add_orct2_test(INI IniWriterTest.cpp IniReaderTest.cpp)
 add_orct2_test(String StringTest.cpp)
 
 if(ENABLE_RCT2_TESTS)
-    execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_LIST_DIR}/testdata" "${CMAKE_CURRENT_BINARY_DIR}/testdata")
+    if(NOT WIN32)
+        execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_LIST_DIR}/testdata" "${CMAKE_CURRENT_BINARY_DIR}/testdata")
```
Comment (janisozaur) [inline test/tests/CMakeLists.txt]: Why is it skipping the symlinking?

Comment (vector-of-bool) [reply test/tests/CMakeLists.txt]: `create_symlink` doesn't work properly on Windows, as far as I've seen. It may be sufficient to just outright copy the directory instead for Windows.

File: src/openrct2-dll/openrct2-dll.cpp [hunk 2]
```diff
@@ -41,7 +41,9 @@ static char * * GetCommandLineArgs(int argc, wchar_t * * argvW);
 static void FreeCommandLineArgs(int argc, char * * argv);
 static char * ConvertUTF16toUTF8(const wchar_t * src);
 
-DLLEXPORT int LaunchOpenRCT2(int argc, wchar_t * * argvW)
+DLLEXPORT int LaunchOpenRCT2(int argc, wchar_t ** argvW);
```
Comment (janisozaur) [inline src/openrct2-dll/openrct2-dll.cpp]: Why is this line here?

File: test/tests/CMakeLists.txt [hunk 3]
```diff
@@ -1,163 +1,39 @@
...
-if (NOT DISABLE_RCT2_TESTS)
-    add_test(NAME ride_ratings COMMAND test_ride_ratings)
-    add_test(NAME multilaunch COMMAND test_multilaunch)
-endif ()
+    set(ENABLE_RCT2_TESTS OFF)
+endif()
+
+function(add_orct2_test name)
+    add_executable(test-${name} ${ARGN})
+    target_link_libraries(test-${name} PRIVATE openrct2::core ${GTEST_LIBRARIES})
```
Comment (janisozaur) [inline test/tests/CMakeLists.txt]: This assumes all tests link to `libopenrct2`, I specifically wanted to avoid that to ensure that parts of the project are standalone.

File: .travis.yml [hunk 4]
```diff
@@ -65,19 +65,19 @@ matrix:
               else curl --progress-bar --upload-file openrct2-linux.tar.gz https://transfer.sh/openrct2-linux-i686.tar.gz -o link && cat link;
               fi
         - os: linux
-          env: OPENRCT2_CMAKE_OPTS="-G Ninja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DBUILD_SHARED_LIBS=ON -DDISABLE_RCT2_TESTS=on" TARGET=ubuntu_amd64
+          env: OPENRCT2_CMAKE_OPTS="-G Ninja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DBUILD_SHARED_LIBS=ON -DENABLE_RCT2_TESTS=NO" TARGET=ubuntu_amd64
         - os: linux
-          env: OPENRCT2_CMAKE_OPTS="-G Ninja -DCMAKE_TOOLCHAIN_FILE=../CMakeLists_mingw.txt -DFORCE32=on -DBUILD_SHARED_LIBS=ON -DDISABLE_RCT2_TESTS=on" TARGET=windows
+          env: OPENRCT2_CMAKE_OPTS="-G Ninja -C../cmake/configs/LinuxMinGW.cmake -DBUILD_SHARED_LIBS=ON -DENABLE_RCT2_TESTS=NO" TARGET=windows
```
Comment (janisozaur) [inline .travis.yml]: is `-C` a shorthand for `-DCMAKE_TOOLCHAIN_FILE=`?

Comment (janisozaur) [reply .travis.yml]: `LinuxMinGW` implies that Linux is required to build it, which is not true. Mingw-w64 can be used from other systems, too.

FWIW: We require Mingw-w64, MinGW is a bit different thing. Their naming is tricky and there is some drama involved.

File: flags.cmake [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        -Wnull-dereference
+        $<$<COMPILE_LANGUAGE:CXX>:-Wsuggest-final-types>
+        $<$<COMPILE_LANGUAGE:CXX>:-Wsuggest-final-methods>
+        $<$<COMPILE_LANGUAGE:CXX>:-Wsuggest-override>
+        $<$<COMPILE_LANGUAGE:CXX>:-Wnon-virtual-dtor>
+        -Wduplicated-cond
+        -Wstrict-overflow=1
+        -Wimplicit
+        )
+    set(maybe_cxx_flags)
```
Comment (janisozaur) [inline flags.cmake]: What's this? I see C++ flags already handled in `maybe_flags` so is this still needed?

File: flags.cmake [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    target_compile_options(openrct2-flags-iface INTERFACE -Wno-error=objc-method-access)
+endif()
+
+# Enable dlopen on Unix systems
+if(UNIX AND NOT CMAKE_SYSTEM_NAME MATCHES BSD)
+    target_link_libraries(openrct2-flags-iface INTERFACE dl)
+endif()
+
+# Enable multithreading
+set(THREADS_PREFER_PTHREAD_FLAG TRUE)
```
Comment (janisozaur) [inline flags.cmake]: So far we only need threads for HTTP request. There are platforms without threading support that OpenRCT2 handles (e.g. emscripten), so this needs to be a compile-time option.

File: test/tests/CMakeLists.txt (hunk 3 above)
Comment (IntelOrca) [reply test/tests/CMakeLists.txt]: I fully support the idea of this on the other hand.

Comment (janisozaur) [reply test/tests/CMakeLists.txt]: That's why you did it so for msbuild and I did the other way for CMake.

File: src/openrct2-dll/openrct2-dll.cpp (hunk 2 above)
Comment (vector-of-bool) [reply src/openrct2-dll/openrct2-dll.cpp]: Was getting warnings about missing decls, but I'm not sure if this is still applicable...

File: .travis.yml (hunk 4 above)
Comment (vector-of-bool) [reply .travis.yml]: `-C` loads up a file that sets initial cache values, so you don't have to specify a ton of `-D` options on the command line. You can give it a good, useful name and move the options for a particular build configuration into that single file.

File: flags.cmake (hunk 5 above)
Comment (vector-of-bool) [reply flags.cmake]: Accident

**Comment by IntelOrca** [comment]
I have now tried this out using VS' cmake support.

**Good:**
* Opening repository in VS and building with CMake works.
* Ninja appears to be faster than msbuild.

**Problems:**
* g2.dat is not built when building `openrct2.exe`, nor is their any target for doing so.
* No manifest created therefore it is not DPI aware. I think the manifest is created dynamically with msbuild using mt.

**Comment by vector-of-bool** [comment]
@IntelOrca, the target would be named `g2`. It doesn't appear? It should be added as an `ALL` default target. What about if you use the Visual Studio generators?

**Comment by IntelOrca** [comment]
1. Open Visual Studio
2. Open folder: openrct2 repo
3. The build dropdowns do not show all and they do not show G2.

File: flags.cmake (hunk 6 above)
Comment (vector-of-bool) [reply flags.cmake]: Fixed

File: src/openrct2/ride/ride.c [hunk 7]
```diff
@@ -4710,8 +4710,12 @@ static void vehicle_create_trains(sint32 rideIndex, sint32 x, sint32 y, sint32 z
     }
 
     // Link the first train and last train together
-    firstTrain.head->prev_vehicle_on_ride = lastTrain.tail->sprite_index;
-    lastTrain.tail->next_vehicle_on_ride = firstTrain.head->sprite_index;
+    if(firstTrain.head != NULL && lastTrain.tail != NULL) {
```
Comment (janisozaur) [inline src/openrct2/ride/ride.c]: This all should be extracted to a separate PR with `openrct2_assert(firstTrain.head != NULL, "Invalid train");` or some other helpful message, for both trains.

Comment (vector-of-bool) [reply src/openrct2/ride/ride.c]: I think this blip can be removed from the PR. It was related to MinGW in release seeing different errors, which was causing Travis errors

Comment (janisozaur) [reply src/openrct2/ride/ride.c]: Yeah, it should still be addressed though. See https://github.com/OpenRCT2/OpenRCT2/pull/6443#issuecomment-336146711 about triggering it in other builds, maybe we can have one of the Arch Travis builds use `-fno-PIC` (it would need `-O2` or `-O3` too)?

File: test/testpaint/CMakeLists.txt [hunk 8]
```diff
@@ -19,8 +19,14 @@ file(GLOB_RECURSE ORCT2_TESTPAINT_SOURCES "${CMAKE_CURRENT_LIST_DIR}/*.c"
...
+    ${ORCT2_RIDE_SOURCES}
+    ${ORCT2_RIDE_DEP_SOURCES}
+    ${ORCT2_TESTPAINT_SOURCES}
+    )
+target_include_directories(testpaint PRIVATE "${PROJECT_SOURCE_DIR}/src/")
+target_compile_definitions(testpaint PRIVATE NO_VEHICLES __TESTPAINT__)
+target_compile_options(testpaint PRIVATE -Wno-unused -fno-omit-frame-pointer)
+target_link_libraries(testpaint PRIVATE openrct2::sections openrct2::flags)
+add_test(NAME testpaint COMMAND testpaint --quiet)
+set_property(TEST testpaint PROPERTY WILL_FAIL TRUE)
```
Comment (janisozaur) [inline test/testpaint/CMakeLists.txt]: The lines above are a bit dark magic, it would be nice to have them commented a little.
* `-fno-omit-frame-pointer` is required for optimised builds to work with RCT2 interop enabled
* `--quiet` is to limit the verbosity of `testpaint`, otherwise it overflows Travis' logs
* `WILL_FAIL` is on, because there are known differences in drawing code. @marijnvdwerf is there an issue about that or other link we could provide for status/explanation?

File: test/tests/CMakeLists.txt [hunk 9]
```diff
@@ -1,163 +1,39 @@
...
+add_orct2_test(SawyerCoding sawyercoding_test.cpp)
+add_orct2_test(LanguagePack LanguagePackTest.cpp)
+add_orct2_test(INI IniWriterTest.cpp IniReaderTest.cpp)
+add_orct2_test(String StringTest.cpp)
+
+if(ENABLE_RCT2_TESTS)
+    if(NOT WIN32)
+        execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_LIST_DIR}/testdata" "${CMAKE_CURRENT_BINARY_DIR}/testdata")
+    endif()
+    add_orct2_test(RideRatings RideRatings.cpp TestData.cpp)
```
Comment (janisozaur) [inline test/tests/CMakeLists.txt]: These tests should be still built, even if not launched on Travis.

File: test/tests/get-gtest.cmake [hunk 10]
```diff
@@ -0,0 +1,65 @@
+# Gets GTest from a remote
+include(ExternalProject)
+
+ExternalProject_Add(
+    googletest-distribution
+    URL https://github.com/google/googletest/archive/release-1.8.0.tar.gz
```
Comment (janisozaur) [inline test/tests/get-gtest.cmake]: Make sure to update the link, as it changed on `develop`: https://github.com/OpenRCT2/OpenRCT2/commit/e01cd0fd40bac48b32a11bad65ee4cad499474cf

File: .travis.yml [hunk 11]
```diff
@@ -22,62 +22,98 @@ matrix:
...
-              else curl --progress-bar --upload-file openrct2-linux.tar.gz https://transfer.sh/openrct2-linux-i686.tar.gz -o link && cat link;
+              if [[ "z$OPENRCT2_ORG_TOKEN" != "z" && "$TRAVIS_PULL_REQUEST" == "false" && ("${TRAVIS_BRANCH}" =~ ^(develop|push/) || "z${TRAVIS_TAG}" != "z") ]]; then
+                curl -o - \
+                  -v \
+                  --form "key=$OPENRCT2_ORG_TOKEN" \
+                  --form "fileName=OpenRCT2-${OPENRCT2_VERSION}${FILENAME_PART}-linux-i686.tar.gz" \
+                  --form "version=${OPENRCT2_VERSION}" \
+                  --form "gitHash=$TRAVIS_COMMIT" \
+                  --form "gitBranch=$PUSH_BRANCH" \
+                  --form "flavourId=9" \
```
Comment (janisozaur) [inline .travis.yml]: 4.

This part is almost the same for amd64 and i686 jobs, maybe you can extract it to a script? We could then reuse it for Android's APKs too.

File: .travis.yml [hunk 12]
```diff
@@ -142,25 +178,34 @@ matrix:
         # Following entries used to be included in testing, but they only proved useful while changing things in CMake setup.
         # They are meant to be used when there are changes to CMakeLists.txt
         # - os: linux
-        #   env: OPENRCT2_CMAKE_OPTS="-DDISABLE_NETWORK=ON -DDISABLE_HTTP_TWITCH=ON -DCMAKE_C_COMPILER=gcc-4.8 -DCMAKE_CXX_COMPILER=g++-4.8"
+        #   env: OPENRCT2_CMAKE_OPTS="-DENABLE_NETWORK=NO -DENABLE_HTTP_TWITCH=NO -DCMAKE_C_COMPILER=gcc-4.8 -DCMAKE_CXX_COMPILER=g++-4.8"
```
Comment (janisozaur) [inline .travis.yml]: These were not tested in a looooong time… They should be tested at least once now or removed. GCC 4.8 is probably not going to fulfil our C++14 needs now.

File: CMakeLists.txt [hunk 13]
```diff
@@ -1,161 +1,74 @@
 # CMAKE project for openrct2
-cmake_minimum_required(VERSION 3.1)
+cmake_minimum_required(VERSION 3.9)
 if (CMAKE_BINARY_DIR STREQUAL CMAKE_SOURCE_DIR)
     message(FATAL_ERROR "Building in-source is not supported! Create a build dir and remove ${CMAKE_SOURCE_DIR}/CMakeCache.txt")
 endif()
 
-project(openrct2)
+project(openrct2 VERSION 0.1.5 LANGUAGES C CXX)
```
Comment (janisozaur) [inline CMakeLists.txt]: Is this version needed? Where is it used? Do we have to take care of it when doing releases?

File: CMakeLists.txt [hunk 14]
```diff
@@ -1,161 +1,74 @@
...
 
-include(FindPkgConfig)
-include(CheckCCompilerFlag)
-include(CheckCXXCompilerFlag)
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")
+
+# Put all runtime binaries in <build>/bin
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/bin")
+# Copy data into the bin dir. This permits running easily out of the build directory.
+file(COPY data DESTINATION bin)
```
Comment (janisozaur) [inline CMakeLists.txt]: When is it getting copied? During configuration (`cmake ../`) or installation (`make install`)? Will it fail if I already have it symlinked?

File: test/testpaint/CMakeLists.txt (hunk 8 above)
Comment (marijnvdwerf) [reply test/testpaint/CMakeLists.txt]: @janisozaur I don't think so. Might be good to create one though.

**Comment by janisozaur** [comment]
@vector-of-bool are you still working on this?

**Comment by vector-of-bool** [comment]
Yes, this is still active. I haven't made any changes yet, but I intend to get back. I'm busy IRL and with another one of my projects. Sorry for the delay.

File: test/testpaint/CMakeLists.txt (hunk 8 above)
Comment (janisozaur) [reply test/testpaint/CMakeLists.txt]: ping @marijnvdwerf

**Comment by janisozaur** [comment]
hi @vector-of-bool, still there?

**Comment by vector-of-bool** [comment]
Yes, I'm still alive. I'm really sorry to go dark on this. I've got a lot of things going on right now, especially with my own OSS project that's pending a big refactor. Time just seems to get away from me. I haven't forgotten about this, I promise.

**Comment by janisozaur** [comment]
Do you think there's any chance of separating this into smaller chunks? I imagine the changes for finding the packages can be done without other parts of this refactor?

**Comment by janisozaur** [comment]
Right now we are cleaning up all open PRs and I'm very sad I need to close this, as it has grown stale. There is lot of improvement in here, so if you ever feel like revisiting this @vector-of-bool (or anyone else) please do so. I think some of proposed changes were already included and we can build using MSVC from CMake too.

**Comment by vector-of-bool** [comment]
@janisozaur all good. This grew stale and was extremely large in scope. I'm currently wrapped up in cmake-tools stuff, and I really want to find the time to get back here. Maybe some of my changes can inspire some smaller and more incremental fixes. Sorry that I went dark on you all.
